### PR TITLE
adding possibility to work with Etags

### PR DESCRIPTION
**Pull request by Alexandru-Dumitru**
- add zosmf header definitions
    - add zosmf Rest Get operation that returns Etag
    - add necessary tests
    - add conditional processing if Etag option is passed
    in Download and Upload APIs
    - add returnEtag as a Download / Get API Option
    - add  etag as an Upload API Option

Signed-off-by: Alexandru-Paul Dumitru <ad670553@broadcom.net>

### CODE REVIEW

**Comment by Alexandru-Dumitru** [comment]
branch test failed because of audit run (was visible on Friday). Now branch is not there to rebuild, because of pipeline setup.
Would resubmit if needed. Thanks.

File: packages/zosfiles/src/api/methods/get/Get.ts [hunk 1]
```diff
@@ -50,10 +50,22 @@ export class Get {
             endpoint = posix.join(ZosFilesConstants.RESOURCE, ZosFilesConstants.RES_DS_FILES, `-(${options.volume})`, dataSetName);
         }
 
+        let dsContent: Buffer;
+        let restResponse: {data: Buffer; etag: string};
```
[the remark this answers is not in the document]
Comment (zFernand0) [reply packages/zosfiles/src/api/methods/get/Get.ts]: you may want to return the signature of this function to indicate the possible return values i.e. `: Promise<Buffer>` needs updated

File: packages/zosfiles/src/api/methods/get/Get.ts [hunk 2]
```diff
@@ -63,11 +75,11 @@ export class Get {
      * @param {string}           USSFileName  - contains the data set name
      * @param {IViewOptions} [options={}] - contains the options to be sent
      *
-     * @returns {Promise<Buffer>} Promise that resolves to the content of the uss file
+     * @returns {Promise<any>} Promise that resolves to the content of the uss file
      *
      * @throws {ImperativeError}
      */
-    public static async USSFile(session: AbstractSession, USSFileName: string, options: IGetOptions = {}): Promise<Buffer> {
+    public static async USSFile(session: AbstractSession, USSFileName: string, options: IGetOptions = {}): Promise<any> {
```
Comment (zFernand0) [inline packages/zosfiles/src/api/methods/get/Get.ts]: Let's try to avoid `any` in the future.

File: packages/zosfiles/src/api/methods/get/Get.ts [hunk 3]
```diff
@@ -87,8 +99,23 @@ export class Get {
                 reqHeaders = [ZosmfHeaders.X_IBM_BINARY];
             }
         }
-        const content = await ZosmfRestClient.getExpectBuffer(session, endpoint, reqHeaders);
 
-        return content;
+        let fileContent: Buffer;
+        let restResponse: {data: Buffer; etag: string};
```
Comment (zFernand0) [inline packages/zosfiles/src/api/methods/get/Get.ts]: 👆

File: packages/zosfiles/src/api/methods/get/Get.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/get/Get.ts]: fixed with https://github.com/zowe/zowe-cli/pull/588/commits/4d836a3c22d6e6bfd7d451d7b162301979a0a9cb

File: packages/zosfiles/src/api/methods/download/Download.ts [hunk 4]
```diff
@@ -237,10 +260,18 @@ export class Download {
...
+                response = restResponse;
+            } else {
+                response = fileContent;
+            }
+
             return {
                 success: true,
                 commandResponse: util.format(ZosFilesMessages.ussFileDownloadedSuccessfully.message, destination),
-                apiResponse: fileContent
+                apiResponse: response
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/download/Download.ts]: agreed offline - not needed :D

File: packages/zosfiles/src/api/methods/get/Get.ts (hunk 3 above)
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/get/Get.ts]: fixed with https://github.com/zowe/zowe-cli/pull/588/commits/4d836a3c22d6e6bfd7d451d7b162301979a0a9cb

File: packages/zosfiles/src/api/methods/get/Get.ts (hunk 2 above)
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/get/Get.ts]: fixed with https://github.com/zowe/zowe-cli/pull/588/commits/4d836a3c22d6e6bfd7d451d7b162301979a0a9cb

File: packages/zosfiles/src/api/methods/download/Download.ts [hunk 5]
```diff
@@ -221,7 +237,14 @@ export class Download {
         ImperativeExpect.toNotBeNullOrUndefined(ussFileName, ZosFilesMessages.missingUSSFileName.message);
         ImperativeExpect.toNotBeEqual(ussFileName, "", ZosFilesMessages.missingUSSFileName.message);
         try {
-            let fileContent = await Get.USSFile(session, ussFileName, {binary: options.binary});
+            let fileContent: Buffer;
+            let restResponse: {data: Buffer; etag: string};
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/download/Download.ts]: added with https://github.com/zowe/zowe-cli/pull/588/commits/4d836a3c22d6e6bfd7d451d7b162301979a0a9cb

File: packages/zosfiles/src/api/methods/download/Download.ts [hunk 6]
```diff
@@ -79,8 +79,17 @@ export class Download {
                 reqHeaders = [ZosmfHeaders.X_IBM_BINARY];
             }
 
+            let datasetContent: Buffer;
+            let restResponse: {data: Buffer; etag: string};
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/download/Download.ts]: added with https://github.com/zowe/zowe-cli/pull/588/commits/4d836a3c22d6e6bfd7d451d7b162301979a0a9cb

**Pull request merged**